File: app/src/main/java/com/github/android/lvrn/lvrnproject/persistent/repository/impl/TrashDependedRepositoryImpl.java
package com.github.android.lvrn.lvrnproject.persistent.repository.impl;

import androidx.annotation.NonNull;

import com.github.android.lvrn.lvrnproject.persistent.repository.TrashDependedRepository;
import com.github.android.lvrn.lvrnproject.util.PaginationArgs;
import com.github.valhallalabs.laverna.persistent.entity.base.TrashDependedEntity;

import java.util.List;

import static com.github.android.lvrn.lvrnproject.persistent.database.LavernaContract.LavernaBaseTable.COLUMN_ID;
import static com.github.android.lvrn.lvrnproject.persistent.database.LavernaContract.ProfileDependedTable.COLUMN_PROFILE_ID;
import static com.github.android.lvrn.lvrnproject.persistent.database.LavernaContract.TrashDependedTable.COLUMN_TRASH;

/**
 * @author Vadim Boitsov <vadimboitsov1@gmail.com>
 */

public abstract class TrashDependedRepositoryImpl<T extends TrashDependedEntity> extends ProfileDependedRepositoryImpl<T> implements TrashDependedRepository<T> {

    private String mTableName;

    public TrashDependedRepositoryImpl(String mTableName) {
        super(mTableName);
        this.mTableName = mTableName;
    }

    @NonNull
    @Override
    public List<T> getByProfile(@NonNull String profileId, @NonNull PaginationArgs paginationArgs) {
        return getByIdCondition(COLUMN_PROFILE_ID, profileId, getTrashClause(false), paginationArgs);
    }

    @NonNull
    @Override
    public List<T> getTrashByProfile(@NonNull String profileId, @NonNull PaginationArgs paginationArgs) {
        return getByIdCondition(COLUMN_PROFILE_ID, profileId, getTrashClause(true), paginationArgs);
    }

    @Override
    public boolean moveToTrash(@NonNull String entityId) {
        String query = "UPDATE " + mTableName
                + " SET "
                + COLUMN_TRASH + "='" + 1 + "'"
                + " WHERE " + COLUMN_ID + "='" + entityId + "'";
        return super.rawUpdateQuery(query);
    }

    @Override
    public boolean restoreFromTrash(@NonNull String entityId) {
        String query = "UPDATE " + mTableName
                + " SET "
                + COLUMN_TRASH + "=" + 0 + "'"
                + " WHERE " + COLUMN_ID + "='" + entityId + "'";
        return super.rawUpdateQuery(query);
    }

    protected String getTrashClause(boolean isTrash) {
        return " AND " + COLUMN_TRASH + "=" + (isTrash ? "1" : "0");
    }
}

File: app/src/main/java/com/github/android/lvrn/lvrnproject/persistent/repository/impl/TrashDependedRepositoryImpl.kt
package com.github.android.lvrn.lvrnproject.persistent.repository.impl

import com.github.android.lvrn.lvrnproject.persistent.database.LavernaContract.*
import com.github.android.lvrn.lvrnproject.persistent.repository.TrashDependedRepository
import com.github.android.lvrn.lvrnproject.util.PaginationArgs
import com.github.valhallalabs.laverna.persistent.entity.base.TrashDependedEntity

abstract class TrashDependedRepositoryImpl<T : TrashDependedEntity>(tableName: String) : ProfileDependedRepositoryImpl<T>(tableName), TrashDependedRepository<T> {

    override fun getByProfile(profileId: String, paginationArgs: PaginationArgs): MutableList<T> {
        return getByIdCondition(ProfileDependedTable.COLUMN_PROFILE_ID, profileId, getTrashClause(false)!!, paginationArgs)
    }

    override fun getTrashByProfile(profileId: String, paginationArgs: PaginationArgs): MutableList<T> {
        return getByIdCondition(ProfileDependedTable.COLUMN_PROFILE_ID, profileId, getTrashClause(true)!!, paginationArgs)
    }

    override fun moveToTrash(entityId: String): Boolean {
        val query = ("UPDATE " + tableName
                + " SET "
                + TrashDependedTable.COLUMN_TRASH + "='" + 1 + "'"
                + " WHERE " + LavernaBaseTable.COLUMN_ID + "='" + entityId + "'")
        return super.rawUpdateQuery(query)
    }

    override fun restoreFromTrash(entityId: String): Boolean {
        val query = ("UPDATE " + tableName
                + " SET "
                + TrashDependedTable.COLUMN_TRASH + "=" + 0
                + " WHERE " + LavernaBaseTable.COLUMN_ID + "='" + entityId + "'")
        return super.rawUpdateQuery(query)
    }

    override fun removeForPermanent(entityId: String): Boolean {
        val query = ("DELETE FROM " + tableName
                + " WHERE " + LavernaBaseTable.COLUMN_ID + "='" + entityId + "'")
        return super.rawUpdateQuery(query)
    }

    protected fun getTrashClause(isTrash: Boolean): String {
        return " AND " + TrashDependedTable.COLUMN_TRASH + "=" + if (isTrash) "1" else "0"
    }
}